rfcs/45-stageless.md

```diff
@@ -0,0 +1,761 @@
+
+While ordering constraints determine *when* a system runs, **run conditions** determine *if* it runs at all. Run conditions consist of systems that have read-only data access and return `bool`.
+A system or set can have any number of conditions, and it will only run if all of them return `true`. 
+If any condition returns `false`, the system (or systems under the set) will be skipped.
+
+Conditions are each evaluated *at most once* during a single pass of a schedule.
+The evaluation occurs just before the system (or the first system encountered that is part of the set of interest) would be run.
+Because of this, results are guaranteed to be up-to-date. The state read by the conditions cannot be modified in an observable way before the guarded system starts.
+
+```rust
+// This is just an ordinary system: timers need to be ticked!
+fn tick_construction_timer(timer: ResMut<ConstructionTimer>, time: Res<Time>){
+    timer.tick(time.delta());
+}
+
+// This function can be used as a condition because it does not mutate data and returns `bool`.
+fn construction_timer_not_finished(timer: Res<ConstructionTimer>) -> bool {
+    !timer.finished()
+}
+
+// You can use queries, event readers and resources with arbitrarily complex internal logic.
+fn too_many_enemies(
+    population_query: Query<(), With<Enemy>>,
+    population_limit: Res<PopulationLimit>,
+) -> bool {
+    population_query.iter().count() > population_limit
+}
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        // You can add functions with read-only system parameters as conditions.
+        .add_set(GameSet::Construction.run_if(construction_timer_not_finished))
+        .add_many(systems![tick_construction_timer, update_construction_progress].in_set(GameSet::Construction))
+        .add_system(system_meltdown_mitigation.run_if(too_many_enemies))
+        // We can use closures for simple one-off conditions, 
+        // which automatically fetch the appropriate data from the `World`.
+        .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+        // The `resource_exists` and `resource_equals` helper functions for generating
+        // simple conditions (well, closures that convert into conditions).
+        .add_system(gravity.run_if(resource_exists(Gravity)))
+        // Conditions can be attached to system sets, so you can skip the whole set.
+        .add_set(GameSet::Physics.run_if(resource_equals(Paused(false))))
+        .run();
+}
+```
+
+### Running scheduled systems
+
+The systems, sets, and scheduling data added to an `App` is stored on the `World` in the `SystemRegistry` resource. This resource is also responsible for generating and maintaining the dependency graphs for the stored system sets.
+
+In order to run a system or system set on the `World`, the system(s) must be extracted from the `SystemRegistry`. (This lets the `SystemRegistry` remain accessible by the running systems).

```

Comment:
I'm curious why it was chosen to have this live *in* the World instead of outside of it. This section doesn't provide much justification as to why this choice was made. It also does not detail how this would interact with subapps and multiple worlds (i.e. the render app).

I think it's also worth specifying how this is done and the assumptions that are being made here. If we're proposing moving large sets of systems in and out of something like a `HashMap<SystemId, SystemDescriptor>` in a Cell-like remove-mutate-replace pattern, that could add a sizable amount of non-trivial overhead.

Reply:
I forgot to respond to this here. Original Discord convo: <https://discord.com/channels/691052431525675048/749335865876021248/978081305692610590>

> This section doesn't provide much justification as to why this choice was made.

I thought the registry should live in the `World` (as a resource or a field with a special system param) and use a `Cell`-like pattern for access because that was a simple ownership model that allowed for runtime schedule modification.

Systems are already "locked" to the `World` used to initialize them (although that may change), so to me it seemed like the existing separation was more of an API choice. We could make `App` have a `HashMap<WorldId, SystemRegistry>` or similar but that would force `World` to store something that can point back to the `App` (or a lifetimed `World`).

> I think it's also worth specifying how this is done and the assumptions that are being made here.

Any data structures that result from topological sorting can be cached, and any vectors we'd need to hold the systems of a set in top-sorted order for execution could be allocated once and cached as well. 

So I think my main assumption is that if the registry isn't being mutated, the only "necessary" overhead would be moving systems back and forth between the registry and the executable schedules. I'm hoping that overhead is negligible or maybe we can just "remove" once and skip the "replace" step for the main schedule while it's looping.

> multi-world

I think multi-world is beyond the scope of this RFC. At the very least, this RFC should not change the current situation.

AFAIK systems have to remain `Send` and `Sync` even if the data accessed may be `!Send`. In such cases it should be the executor's responsibility to run those systems on the required thread, just like how it's currently responsible for ensuring non-conflicting borrows.